model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/connections/DefaultHotRodConnectionProviderFactory.java

```diff
@@ -111,32 +118,53 @@ public void lazyInit() {
 
         remoteBuilder.addContextInitializer(CommonPrimitivesProtoSchemaInitializer.INSTANCE);
         ENTITY_DESCRIPTOR_MAP.values().stream().map(HotRodEntityDescriptor::getProtoSchema).forEach(remoteBuilder::addContextInitializer);
-        remoteCacheManager = new RemoteCacheManager(remoteBuilder.build());
 
-        Set<String> remoteCaches = ENTITY_DESCRIPTOR_MAP.values().stream()
-                .map(HotRodEntityDescriptor::getCacheName).collect(Collectors.toSet());
+        // Configure settings necessary for locking
+        configureLocking(remoteBuilder);
 
-        LOG.debugf("Uploading proto schema to Infinispan server.");
-        registerSchemata();
+        remoteCacheManager = new RemoteCacheManager(remoteBuilder.build());
 
-        String reindexCaches = config.get("reindexCaches", null);
-        RemoteCacheManagerAdmin administration = remoteCacheManager.administration();
-        if (reindexCaches != null && reindexCaches.equals("all")) {
-            LOG.infof("Reindexing all caches. This can take a long time to complete. While the rebuild operation is in progress, queries might return fewer results.");
-            remoteCaches.stream()
-                    .peek(remoteCacheManager::getCache) // access the caches to force their creation, otherwise reindexing fails if cache doesn't exist
-                    .forEach(administration::reindexCache);
-        } else if (reindexCaches != null && !reindexCaches.isEmpty()){
-            Arrays.stream(reindexCaches.split(","))
-                .map(String::trim)
-                    .filter(e -> !e.isEmpty())
-                    .filter(remoteCaches::contains)
-                    .peek(cacheName -> LOG.infof("Reindexing %s cache. This can take a long time to complete. While the rebuild operation is in progress, queries might return fewer results.", cacheName))
-                    .peek(remoteCacheManager::getCache) // access the caches to force their creation, otherwise reindexing fails if cache doesn't exist
-                    .forEach(administration::reindexCache);
+        // Acquire initial phase lock to avoid concurrent schema update
+        RemoteCache<String, String> locksCache = remoteCacheManager.getCache(HOT_ROD_LOCKS_CACHE_NAME);
+        try {
```

Comment:
I think this would result in an infinite loop as `HotRodLockProvider` requires `HotRodConnectionProvider`.

Reply:
Thank you for pointing me to the fact that there's a cyclical dependency between those two provider. In my case it worked because `HotRodLockProvider` was requested after initialization of `remoteCacheManager` which means `lazyInit()` from `DefaultHotRodConnectionProviderFactory` isn't executed for the second time.

Reply:
Well, this is a `chicken-and-egg` problem. I see that at the point of the locking, the `remoteCacheManager` is already initialized, so the `HotRodGlobalLockProvider` theoretically should work when requesting only `locks` cache. Other caches would not work as schemata are not yet initialized. However, I would prefer not to call the `DefaultHotRodConnectionProvider.create` method until the factory does not fully finish the `lazyInit` method. WDYT?

Reply:
Yeah, I agree.